xla/backends/cpu/runtime/onednn/onednn_op_thunk.cc

```diff
@@ -0,0 +1,192 @@
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "absl/base/call_once.h"
+#include "absl/container/inlined_vector.h"
+#include "absl/functional/function_ref.h"
+#include "absl/log/check.h"
+#include "absl/memory/memory.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/str_format.h"
+#include "absl/types/span.h"
+#include "xla/backends/cpu/runtime/onednn/onednn_threadpool.h"
+#include "xla/backends/cpu/runtime/thunk.h"
+#include "xla/runtime/buffer_use.h"
+#include "xla/service/cpu/onednn_matmul.h"
+#include "xla/status_macros.h"
+#include "xla/stream_executor/device_memory.h"
+#include "xla/tsl/concurrency/async_value_ref.h"
+#include "xla/tsl/platform/logging.h"
+#include "xla/tsl/platform/statusor.h"
+
+namespace xla::cpu {
+
+// oneDNN runtime instantiated for the oneDNN operation.
+struct OneDnnOpThunk::OneDnnRuntime {
+  OneDnnRuntime(Eigen::ThreadPoolInterface* thread_pool);
+
+  OneDnnRuntime(OneDnnRuntime&&) = default;
+  OneDnnRuntime& operator=(OneDnnRuntime&&) = default;
+
+  tsl::AsyncValueRef<OneDnnOpThunk::ExecuteEvent> Invoke(
+      Eigen::ThreadPoolInterface* thread_pool,
+      absl::Span<MemrefInfoHandler> arguments,
+      absl::Span<MemrefInfoHandler> results,
+      const OneDnnOpThunk::OneDnnOpConfig& config, const std::string& target);
+
+  std::unique_ptr<OneDnnThreadPool> threadpool;
+
+  dnnl::engine cpu_engine;
+  dnnl::stream onednn_stream;
+  // We initialize the resources struct here to default values, so that we can
+  // keep the primitive and memory objects alive for the duration of the
+  // runtime. Otherwise, they would be destroyed as soon as we exit the
+  // ExecuteOneDnn<primitive> method. This is a requirement of
+  // oneDNN library's asynchronous execution model.
+  OneDnnResources resources;
+};
+
+OneDnnOpThunk::OneDnnRuntime::OneDnnRuntime(
+    Eigen::ThreadPoolInterface* thread_pool)
+    : threadpool(
+          std::make_unique<OneDnnThreadPool>(thread_pool, /*is_async=*/true)),
+      cpu_engine(dnnl::engine::kind::cpu, 0),
+      onednn_stream(
+          dnnl::threadpool_interop::make_stream(cpu_engine, threadpool.get())),
+      resources() {}
+
+tsl::AsyncValueRef<OneDnnOpThunk::ExecuteEvent>
+OneDnnOpThunk::OneDnnRuntime::Invoke(
+    Eigen::ThreadPoolInterface* thread_pool,
+    absl::Span<MemrefInfoHandler> arguments,
+    absl::Span<MemrefInfoHandler> results,
+    const OneDnnOpThunk::OneDnnOpConfig& config, const std::string& target) {
+  // Update threadpool
+  threadpool->set_thread_pool(thread_pool);
+
+  // TODO(intel-tf): Add support for more oneDNN operations as needed.
+  static absl::once_flag log_once_flag;
+  absl::call_once(log_once_flag, [&] {
+    VLOG(0) << absl::StreamFormat(
+        "Executing oneDNN thunk with target `%s`: num_args=%d, num_results=%d",
+        target, arguments.size(), results.size());
+  });
+
+  if (target == "__onednn$matmul") {
+    const auto& matmul_config = absl::get<OneDnnMatMulConfig>(config);
```

Comment:
For this function, we prefer the standard version.
```suggestion
    const auto& matmul_config = std::get<OneDnnMatMulConfig>(config);
```